pallets/lbp/src/lib.rs

```diff
@@ -0,0 +1,1017 @@
+				Error::<T>::SaleStarted
+			);
+
+			if let Some(new_start) = start {
+				pool_data.start = new_start;
+			}
+
+			if let Some(new_end) = end {
+				pool_data.end = new_end;
+			}
+
+			if let Some((w1, w2)) = initial_weights {
+				pool_data.initial_weights = (w1, w2);
+			}
+
+			if let Some((w1, w2)) = final_weights {
+				pool_data.final_weights = (w1, w2);
+			}
+
+			Self::validate_pool_data(&pool_data)?;
+
+			<PoolData<T>>::insert(&pool_id, &pool_data);
+			Self::deposit_event(Event::PoolUpdated(who, pool_id));
+
+			Ok(().into())
+		}
+
+		#[pallet::weight(<T as Config>::WeightInfo::pause_pool())]
+		pub fn pause_pool(origin: OriginFor<T>, pool_id: PoolId<T>) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+
+			ensure!(<PoolOwner<T>>::contains_key(&pool_id), Error::<T>::PoolNotFound);
+
+			Self::ensure_pool_ownership(&who, &pool_id)?;
+
+			let mut pool_data = Self::pool_data(&pool_id);
+
+			ensure!(pool_data.pausable, Error::<T>::PoolIsNotPausable);
+
+			ensure!(!pool_data.paused, Error::<T>::CannotPausePausedPool);
+
+			let now = <frame_system::Pallet<T>>::block_number();
+			ensure!(pool_data.end > now, Error::<T>::CannotPauseEndedPool);
+
+			pool_data.paused = true;
+			<PoolData<T>>::insert(&pool_id, &pool_data);
+
+			Self::deposit_event(Event::Paused(who, pool_id));
+			Ok(().into())
+		}
+
+		#[pallet::weight(<T as Config>::WeightInfo::unpause_pool())]
+		pub fn unpause_pool(origin: OriginFor<T>, pool_id: PoolId<T>) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+
+			ensure!(<PoolOwner<T>>::contains_key(&pool_id), Error::<T>::PoolNotFound);
+
+			Self::ensure_pool_ownership(&who, &pool_id)?;
+
+			let mut pool_data = Self::pool_data(&pool_id);
+
+			ensure!(pool_data.paused, Error::<T>::PoolIsNotPaused);
+
+			let now = <frame_system::Pallet<T>>::block_number();
+			ensure!(pool_data.end > now, Error::<T>::CannotUnpauseEndedPool);
+
+			pool_data.paused = false;
+			<PoolData<T>>::insert(&pool_id, &pool_data);
+
+			Self::deposit_event(Event::Unpaused(who, pool_id));
+			Ok(().into())
+		}
+
+		#[pallet::weight(<T as Config>::WeightInfo::add_liquidity())]
+		#[transactional]
+		pub fn add_liquidity(
+			origin: OriginFor<T>,
+			pool_id: PoolId<T>,
+			amount_a: BalanceOf<T>,
+			amount_b: BalanceOf<T>,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+
+			ensure!(<PoolOwner<T>>::contains_key(&pool_id), Error::<T>::PoolNotFound);
+
+			Self::ensure_pool_ownership(&who, &pool_id)?;
+
+			ensure!(
+				!amount_a.is_zero() || !amount_b.is_zero(),
+				Error::<T>::CannotAddZeroLiquidity
+			);
+
+			let (mut reserve_a, mut reserve_b) = Self::pool_balances(&pool_id);
+			let (asset_a, asset_b) = Self::pool_assets(&pool_id);
+
+			let pool_data = Self::pool_data(&pool_id);
+
+			let now = <frame_system::Pallet<T>>::block_number();
+			ensure!(
+				pool_data.start == Zero::zero() || now < pool_data.start,
+				Error::<T>::SaleStarted
+			);
```

Comment:
You can move this before reading balances from the storage. So in case of error - storage calls to get balances and assets are not executed .